cmd/dnn-ssd/main.go

```diff
@@ -0,0 +1,140 @@
+// (no fear, this one is fairly small, 5.1mb)
+//
+// Also, you will need the prototxt file:
+// https://raw.githubusercontent.com/opencv/opencv/master/samples/dnn/face_detector/deploy.prototxt
+//
+// How to run:
+//
+// 		go run ./cmd/dnn-ssd/main.go 0 [protofile] [modelfile]
+//
+// +build example
+
+package main
+
+import (
+	"fmt"
+	"image"
+	"image/color"
+	"os"
+	"strconv"
+
+	"gocv.io/x/gocv"
+)
+
+func min(a, b float32) float32 {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func max(a, b float32) float32 {
+	if a > b {
+		return a
+	}
+	return b
+}
+
+func main() {
+	if len(os.Args) < 4 {
+		fmt.Println("How to run:\ndnn-ssd [camera ID] [protofile] [modelfile]")
+		return
+	}
+
+	// parse args
+	deviceID, _ := strconv.Atoi(os.Args[1])
+	proto := os.Args[2]
+	model := os.Args[3]
+
+	// open capture device
+	webcam, err := gocv.VideoCaptureDevice(int(deviceID))
+	if err != nil {
+		fmt.Printf("Error opening video capture device: %v\n", deviceID)
+		return
+	}
+	defer webcam.Close()
+
+	window := gocv.NewWindow("Caffe Classifier")
+	defer window.Close()
+
+	img := gocv.NewMat()
+	defer img.Close()
+
+	// open DNN classifier
+	net := gocv.ReadNetFromCaffe(proto, model)
+	if net.Empty() {
+		fmt.Printf("Error reading network model from : %v %v\n", proto, model)
+		return
+	}
+	defer net.Close()
+
+	green := color.RGBA{0, 255, 0, 0}
+	fmt.Printf("Start reading camera device: %v\n", deviceID)
+
+	for {
+		if ok := webcam.Read(img); !ok {
+			fmt.Printf("Error cannot read device %d\n", deviceID)
+			return
+		}
+		if img.Empty() {
+			continue
+		}
+
+		W := float32(img.Cols())
+		H := float32(img.Rows())
+
+		// convert image Mat to 96x128 blob that the detector can analyze
+		blob := gocv.BlobFromImage(img, 1.0, image.Pt(128, 96), gocv.NewScalar(104.0, 177.0, 123.0, 0), false, false)
+
+		// feed the blob into the classifier
+		net.SetInput(blob, "data")
+
+		// run a forward pass thru the network
+		detBlob := net.Forward("detection_out")
+
+		// extract the detections (nfaces x 7features)
+		detections := gocv.GetBlobChannel(detBlob, 0, 0)
+
+		// each detection is on a row
+		for r := 0; r < detections.Rows(); r++ {
+			// you would want this for general object detection, but not needed here.
+			// classid := detections.GetFloatAt(r, 1)
+
+			confidence := detections.GetFloatAt(r, 2)
+			if confidence < 0.5 {
+				continue
+			}
+
+			left := detections.GetFloatAt(r, 3) * W
+			top := detections.GetFloatAt(r, 4) * H
+			right := detections.GetFloatAt(r, 5) * W
+			bottom := detections.GetFloatAt(r, 6) * H
+
+			// scale to video size:
+			left = min(max(0, left), W-1)
+			right = min(max(0, right), W-1)
+			bottom = min(max(0, bottom), H-1)
+			top = min(max(0, top), H-1)
+
+			// draw it
+			rect := image.Rect(int(left), int(top), int(right), int(bottom))
+			gocv.Rectangle(img, rect, green, 3)
+		}
+
+		blob.Close()
```

Comment:
shouldn't we `defer` the `Close()` calls earlier?

Reply:
ok.